src/lib.rs

```diff
@@ -0,0 +1,238 @@
+
+/// Atomic execution coordinator actor methods available
+#[derive(FromPrimitive)]
+#[repr(u64)]
+pub enum Method {
+    Constructor = METHOD_CONSTRUCTOR,
+    PreCommit = 2,
+    Revoke = 3,
+}
+
+/// Atomic execution coordinator actor
+pub struct Actor;
+
+impl Actor {
+    fn constructor<BS, RT>(rt: &mut RT, params: ConstructorParams) -> Result<(), ActorError>
+    where
+        BS: Blockstore,
+        RT: Runtime<BS>,
+    {
+        rt.validate_immediate_caller_is(std::iter::once(&*INIT_ACTOR_ADDR))?;
+
+        let st = State::new(rt.store(), params).map_err(|e| {
+            e.downcast_default(ExitCode::USR_ILLEGAL_STATE, "Failed to create actor state")
+        })?;
+        rt.create(&st)?;
+        Ok(())
+    }
+
+    /// Records a pre-commitment from an actor to perform an atomic
+    /// execution. This method is to be invoked by a wrapped crossnet
+    /// message originating in one of the execution actors involved in
+    /// the atomic execution. Once the coordinator actor collects
+    /// pre-commitments from all the execution actors, it emits for
+    /// each of the execution actors a crossnet message triggering the
+    /// specified method to commit the atomic execution.
+    fn pre_commit<BS, RT>(rt: &mut RT, params: ApplyMsgParams) -> Result<bool, ActorError>
+    where
+        BS: Blockstore,
+        RT: Runtime<BS>,
+    {
+        let st: State = rt.state()?;
+        rt.validate_immediate_caller_is(std::iter::once(&st.ipc_gateway_address))?;
+
+        let ApplyMsgParams {
+            cross_msg:
+                CrossMsg {
+                    msg: StorableMsg { from, params, .. },
+                    ..
+                },
+        } = params;
+
+        let params: PreCommitParams = cbor::deserialize_params(&params)?;
+        let actors = &params.actors;
+        let exec_id = &params.exec_id;
+
+        if !actors.contains(&from) {
+            return Err(actor_error!(
+                illegal_argument,
+                "unexpected cross-message origin"
+            ));
+        }
+
+        let msgs = rt.transaction(|st: &mut State, rt| {
+            st.modify_atomic_exec(rt.store(), exec_id.into(), actors.clone(), |entry| {
+                // Record the pre-commitment
+                entry.set(BytesKey::from(from.to_bytes()?), params.commit)?;
+
+                // Check if any pre-commitment is missing
+                for actor in actors {
+                    let k = BytesKey::from(actor.to_bytes()?);
+                    if !entry.contains_key(&k)? {
+                        return Ok(None);
+                    }
+                }
+
+                // Prepare messages to commit the atomic execution
+                let mut msgs = Vec::new();
+                entry.for_each(|addr, &method| {
+                    msgs.push(CrossMsg {
+                        msg: StorableMsg {
+                            to: IPCAddress::from_bytes(addr)?,
+                            method,
```

Comment:
Shouldn't this be a specific entrypoint to dispatch atomic exec commitments (I don't know if calling the method directly would work)?

Reply:
Not sure I got your question exactly. Do you mean reserving fixed method numbers in execution actors to commit/rollback the atomic execution? 

So far I wanted to give the execution actors flexibility and let them specify the callback method number themselves, but we may revise this once [FIP-0042](https://github.com/filecoin-project/FIPs/blob/master/FRCs/frc-0042.md) is finalized.

Reply:
Can you remind me what are the stages of the execution? There at least two methods for each contract involved in the protocol: one for the messages involved in the atomic execution, and one to commit the final state. How I imagine the end-to-end is that:
- A set of actors decide to perform an atomic execution that is represented through a set of messages sent to the actors in different subnets.
- Once this execution is performed and the coordinator actors decides it has been done successfully, the corresponding outputs need to be committed in the state of the actors in each subnet, but what is the method responsible for accepting this commitment message and merging the state, the same as the ones called in the atomic exec?

Reply:
The happy path roughly looks like follows (more details [here](https://github.com/protocol/ConsensusLab/discussions/154#discussioncomment-3882477)).

- First, each of the execution actors locally initializes a new instance of the atomic execution protocol by locking the relevant potion of its state and producing a unique atomic execution input (`exec_actor.init_atomic_exec(...)`). 
- Next, the users exchange and collect the full set of inputs, which they use to invoke their execution actors for pre-commitment (`exec_actor.prep_atomic_exec(inputs, ...)`). At this stage, each of the execution actors computes the atomic execution ID, produces and tentatively stores some output, then emits a cross-net message to the coordinator actor (`coord_actor.pre_commit(actors, exec_id, commit)`).
- Once the coordinator actor has recorded a pre-commitment from each of the execution actors, it emits a cross-net message for each of the execution actor to commit the atomic execution (`exec_actor.commit(exec_id)`).
- Finally, each of the execution actor unlocks its state and merges the output once received the commit message from the coordinator actor.